include/ginkgo/core/matrix/fbcsr.hpp

```diff
@@ -0,0 +1,390 @@
+#include <ginkgo/core/base/math.hpp>
+
+
+namespace gko {
+namespace matrix {
+
+
+template <typename ValueType>
+class Dense;
+
+template <typename ValueType, typename IndexType>
+class Csr;
+
+template <typename ValueType, typename IndexType>
+class SparsityCsr;
+
+template <typename ValueType, typename IndexType>
+class Fbcsr;
+
+template <typename ValueType, typename IndexType>
+class FbcsrBuilder;
+
+
+/**
+ * FBCSR is a matrix format meant for matrices having a natural block structure
+ * made up of small, dense, disjoint blocks. It is similar to CSR \sa Csr.
+ * However, unlike Csr, each non-zero location stores a small dense block of
+ * entries having a constant size. This reduces the number of integers that need
+ * to be stored in order to refer to a given non-zero entry, and enables
+ * efficient implementation of certain block methods.
+ *
+ * The block size is expected to be known in advance and passed to the
+ * constructor.
+ *
+ * @note The total number of rows and the number of columns are expected to be
+ *   divisible by the block size.
+ *
+ * The nonzero elements are stored in a 1D array row-wise, and accompanied
+ * with a row pointer array which stores the starting index of each block-row.
+ * An additional block-column index array is used to identify the block-column
+ * of each nonzero block.
+ *
+ * The Fbcsr LinOp supports different operations:
+ *
+ * ```cpp
+ * matrix::Fbcsr *A, *B, *C;      // matrices
+ * matrix::Dense *b, *x;        // vectors tall-and-skinny matrices
+ * matrix::Dense *alpha, *beta; // scalars of dimension 1x1
+ * matrix::Identity *I;         // identity matrix
+ *
+ * // Applying to Dense matrices computes an SpMV/SpMM product
+ * A->apply(b, x)              // x = A*b
+ * A->apply(alpha, b, beta, x) // x = alpha*A*b + beta*x
+ *
+ * // Applying to Fbcsr matrices computes a SpGEMM product of two sparse
+ * matrices A->apply(B, C)              // C = A*B A->apply(alpha, B, beta, C)
+ * // C = alpha*A*B + beta*C
+ *
+ * // Applying to an Identity matrix computes a SpGEAM sparse matrix addition
+ * A->apply(alpha, I, beta, B) // B = alpha*A + beta*B
+ * ```
+ * Both the SpGEMM and SpGEAM operation require the input matrices to be sorted
+ * by block-column index, otherwise the algorithms will produce incorrect
+ * results.
+ *
+ * @tparam ValueType  precision of matrix elements
+ * @tparam IndexType  precision of matrix indexes
+ *
+ * @ingroup fbcsr
+ * @ingroup mat_formats
+ * @ingroup LinOp
+ */
+template <typename ValueType = default_precision, typename IndexType = int32>
+class Fbcsr : public EnableLinOp<Fbcsr<ValueType, IndexType>>,
+              public EnableCreateMethod<Fbcsr<ValueType, IndexType>>,
+              public ConvertibleTo<Fbcsr<next_precision<ValueType>, IndexType>>,
+              public ConvertibleTo<Dense<ValueType>>,
+              public ConvertibleTo<Csr<ValueType, IndexType>>,
+              public ConvertibleTo<SparsityCsr<ValueType, IndexType>>,
```

Comment:
I think `SparsityCsr` should be fine, because the structure of `row_ptrs` and `col_idxs` is the same for both `Csr` and `Fbcsr`. Do you anticipate any issues?

Reply:
I am kind of torn here: On the one hand, having this conversion helps setting up things like Rcm reordering transparently, otherwise, the permutation interface uses raw indices instead of block indices by default, so this creates a bit of inconsistency, unless we create something like BlockPermutable in addition to Permutable. But that would be complex and introduce another issue with variable block sizes. The reason we can't right now directly use block indices is the Permutation LinOp: Its internal storage would then refer to blocks, but its global size would refer to raw rows/columns for LinOp::apply consistency.

Reply:
I was thinking of one interesting option: have a concept of "ring", and let the value_type be any ring. Then, define a full-featured `FixedBlock` with all the operations needed. The only conceptual issue would be that we'd have to take care of multiplication not being commutative. This would make a lot of `Fbcsr` things trivial, in principle. In fact, in principle, we could just have eg. `Csr<FixedBlock<double>>,int>` and just write extra kernels specialized to the case where `value_type` is a `FixedBlock`. This whole thing may not help us write efficient GPU code, but may not necessarily hurt that aspect either. Just a crazy idea - not sure what the implications would be for end-user friendliness or maintainability.

Reply:
I think it is a nice idea, but we would probably break some basic assumptions of LinOp in the process (commutativity as you mention). But maybe we should ponder more on this. I think you should create an issue for now and we can discuss it in the future. 

For now, I think it is okay to go with the conversion to `SparsityCsr` to make sure permutations and reordering works. I am hoping that conversions are reasonably fast.